src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/conditions/ElectronicMonitoringProgrammeService.kt

```diff
@@ -0,0 +1,119 @@
+        "ElectronicMonitoringProvider is null for HdcLicence: $licenceId"
+      }
+
+      else -> error("Trying to update electronic monitoring provider details for non-crd or non-hdc: $licenceId")
+    }
+    electronicMonitoringProvider.isToBeTaggedForProgramme = request.isToBeTaggedForProgramme
+    electronicMonitoringProvider.programmeName = request.programmeName
+    licenceRepository.saveAndFlush(licenceEntity)
+
+    auditService.recordAuditEventUpdateElectronicMonitoringProgramme(licenceEntity, request, staffMember)
+  }
+
+  @Transactional
+  fun handleElectronicMonitoringResponseRecords(licenceEntity: Licence) {
+    val additionalApConditions = licenceEntity.additionalConditions.filter { it.conditionType == "AP" }
+    val conditionsRequiringElectronicMonitoringResponse = licencePolicyService.getConditionsRequiringElectronicMonitoringResponse(licenceEntity.version!!, additionalApConditions.map { it.conditionCode })
+    if (conditionsRequiringElectronicMonitoringResponse.isNotEmpty()) {
+      LicenceConditionService.log.info("Handling Electronic Monitoring response record for conditions: ${conditionsRequiringElectronicMonitoringResponse.joinToString(",") { it.code }}")
+      initialiseElectronicMonitoringProviderIfNotExists(licenceEntity)
+    } else {
+      clearElectronicMonitoringProvider(licenceEntity)
+    }
+  }
+
+  @Transactional
+  fun clearElectronicMonitoringProvider(licenceEntity: Licence) {
+    LicenceConditionService.log.info("Clearing Electronic Monitoring response records for licence: ${licenceEntity.id}")
+    when (licenceEntity) {
+      is CrdLicence -> {
+        licenceEntity.electronicMonitoringProvider?.let {
+          electronicMonitoringProviderRepository.delete(it)
+        }
+        licenceEntity.electronicMonitoringProvider = null
+      }
+      is HdcLicence -> {
+        licenceEntity.electronicMonitoringProvider?.let {
+          electronicMonitoringProviderRepository.delete(it)
+        }
+        licenceEntity.electronicMonitoringProvider = null
+      }
+    }
+  }
+
+  @Transactional
+  fun initialiseElectronicMonitoringProviderIfNotExists(licenceEntity: Licence) {
+    when (licenceEntity) {
+      is CrdLicence -> {
+        if (licenceEntity.electronicMonitoringProvider == null) {
+          licenceEntity.electronicMonitoringProvider = electronicMonitoringProviderRepository.saveAndFlush(
+            ElectronicMonitoringProvider(
+              licence = licenceEntity,
+              isToBeTaggedForProgramme = null,
+              programmeName = null,
+            ),
+          )
+        }
+      }
+      is HdcLicence -> {
+        if (licenceEntity.electronicMonitoringProvider == null) {
+          licenceEntity.electronicMonitoringProvider = electronicMonitoringProviderRepository.saveAndFlush(
+            ElectronicMonitoringProvider(
+              licence = licenceEntity,
+              isToBeTaggedForProgramme = null,
+              programmeName = null,
+            ),
```

Comment:
These two peaces of code are doing the same thing

```
ElectronicMonitoringProvider(
              licence = licenceEntity,
              isToBeTaggedForProgramme = null,
              programmeName = null,
            )
```

Reply:
wondering if we can get rid of `electronicMonitoringProviderRepository` entirely?
and just do: `licenceRepository.save(licence)`

Reply:
May be it was this Merge branch 'main' into [CVSL-2935-Detecting-need-to-add-or-remove-re…](https://dsdmoj.atlassian.net/browse/CVSL-2935-Detecting-need-to-add-or-remove-response)
[…sponse](https://dsdmoj.atlassian.net/browse/CVSL-2935-Detecting-need-to-add-or-remove-response)